src/shredder/reed_solomon.rs

```diff
@@ -33,127 +33,167 @@ pub(super) enum ReedSolomonDeshredError {
+        let decoder = ReedSolomonDecoder::new(DATA_SHREDS, num_coding, MAX_DATA_PER_SHRED).unwrap();
+
+        ReedSolomonCoder {
+            num_coding,
+            encoder,
+            decoder,
+        }
     }
 
-    let coding_offset = TOTAL_SHREDS - num_coding;
+    /// Reed-Solomon encodes the `payload` into [`RawShreds`].
+    ///
+    /// For this, it splits the given slice into [`DATA_SHREDS`] data shreds.
+    /// Then, it generates and adds `num_coding` additional Reed-Solomon coding shreds.
+    ///
+    /// First, however, padding is added to the payload to make it a multiple of `2 * DATA_SHREDS`.
+    /// Bit padding of one 1bit and as many 0 bits as needed is added.
+    /// In the byte representation this looks like `[0x80, 0x00, ..., 0x00]`.
+    ///
+    /// Errors
+    ///
+    /// If the provided payload is larger than [`MAX_DATA_PER_SLICE_AFTER_PADDING`] then returns [`ReedSolomonDeshredError::TooMuchData`].
+    pub(super) fn shred(&mut self, payload: &[u8]) -> Result<RawShreds, ReedSolomonShredError> {
+        if payload.len() > MAX_DATA_PER_SLICE {
+            return Err(ReedSolomonShredError::TooMuchData);
+        }
 
-    // filter to split data and coding shreds
-    let data = shreds.iter().take(coding_offset).filter_map(|s| {
-        s.as_ref().map(|s| match &s.payload_type {
-            ShredPayloadType::Data(d) => (*d.shred_index, &d.data),
-            ShredPayloadType::Coding(_) => panic!("should be a data shred"),
-        })
-    });
+        // determine padding length & configure encoder for shred length
+        let padding_bytes = 2 * DATA_SHREDS - payload.len() % (2 * DATA_SHREDS);
+        let shred_bytes = (payload.len() + padding_bytes).div_ceil(DATA_SHREDS);
+        self.encoder
+            .reset(DATA_SHREDS, self.num_coding, shred_bytes)
+            .unwrap();
 
-    let coding = shreds.iter().skip(coding_offset).filter_map(|s| {
-        s.as_ref().map(|s| match &s.payload_type {
-            ShredPayloadType::Coding(c) => (*c.shred_index - coding_offset, &c.data),
-            ShredPayloadType::Data(_) => panic!("should be a coding shred"),
-        })
-    });
+        // add padding to last shreds
+        let last_shreds_bytes = (2 * DATA_SHREDS).next_multiple_of(shred_bytes);
+        let boundary = payload.len() - (last_shreds_bytes - padding_bytes);
+        let mut last_shreds = Vec::with_capacity(last_shreds_bytes);
+        last_shreds.extend_from_slice(&payload[boundary..]);
+        last_shreds.push(0x80);
+        last_shreds.resize(last_shreds_bytes, 0);
 
-    let restored = rs::decode(DATA_SHREDS, num_coding, data.clone(), coding).unwrap();
+        // chunk data & add coding
+        let mut data = Vec::with_capacity(DATA_SHREDS);
+        payload[..boundary]
+            .chunks(shred_bytes)
+            .chain(last_shreds.chunks(shred_bytes))
+            .for_each(|chunk| {
+                self.encoder.add_original_shard(chunk).unwrap();
```

Comment:
nit: maybe we should add comments about why the unwraps are safe when we interact with the RS crate here and in other places in this module.


I find the error handling the RS crate quite disappointing.  Everything returns the same `Error` enum and one has to look through the call tree to figure out what errors can actually happen and if they are safe to ignore.